pochivm/api_base.h

```diff
@@ -117,111 +117,223 @@ Value<T>::operator Value<U>() const
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MUL>::value> >
-Value<T> operator*(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::MUL>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MUL>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator*(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::MUL, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstTypeHelper::BinaryOps::MUL>(lhs, rhs, AstArithmeticExprType::MUL);
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::DIV>::value> >
-Value<T> operator/(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::DIV>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::DIV>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator/(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::DIV, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstTypeHelper::BinaryOps::DIV>(lhs, rhs, AstArithmeticExprType::DIV);
 }
 
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
-Value<T> operator%(const Value<T>& lhs, const Value<T>& rhs)
+template <typename T, typename U,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<U, AstTypeHelper::BinaryOps::MODULO>::value>,
+          typename = std::enable_if_t<AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::MODULO>::value> >
+Value<typename AstTypeHelper::ArithReturnType<T, U>::type> operator%(const Value<T>& lhs, const Value<U>& rhs)
 {
-    return Value<T>(new AstArithmeticExpr(AstArithmeticExprType::MOD, lhs.__pochivm_value_ptr, rhs.__pochivm_value_ptr));
+    return DoArith<AstTypeHelper::BinaryOps::MODULO>(lhs, rhs, AstArithmeticExprType::MOD);
 }
 
 // Convenience overloading: arithmetic operation with literal
 //
-template<typename T, typename = std::enable_if_t<
-             AstTypeHelper::primitive_type_supports_binary_op<T, AstTypeHelper::BinaryOps::ADD>::value> >
-Value<T> operator+(const Value<T>& lhs, T rhs)
+
+// Do arithmetic of form runtime value OP literal
+//
+template <AstTypeHelper::BinaryOps Binop, typename T, typename U,
```

Comment:
can you reuse the logic of DoArith, so that you write something like `return DoArith(lhs, Literal<U>(rhs), expr_type)`? 
I didn't try myself so let me know if you think this approach could not work.